### ISSUE TITLE
`VisualDensity.adaptivePlatformDensity` should be based on `ThemeData.platform` instead of `defaultTargetPlatform`

### ISSUE THREAD

**Issue by Calpoog** [flutter/flutter#123773]
I'm working on a design system for a Flutter app. This design system supports Android and iOS only. Our component-housing of choice is Widgetbook. Widgetbook, along with many other ways of sharing design systems, is web based. Because of this, Platform reports itself as web. Alternatively, building Widgetbook as a desktop app reports the platform as macOS (or windows/linux). I'd like to see my widgets as if they are on the Android/iOS platform, regardless of true platform. This for some widgets is accomplished by using `Theme.of(context).platform` as specified in the documentation for `defaultTargetPlatform` [here](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/foundation/platform.dart).

"Widgets from the material library should use [Theme.of] to determine the current platform for styling purposes, rather than using [defaultTargetPlatform]"

However, `VisualDensity.adaptivePlatformDensity` which is the default setting for Material button densities uses a switch case on `defaultTargetPlatform` instead of the ambient `ThemeData`s `platform`. This means that without explicitly setting the `visualDensity` to `comfortable` in the individual button's styles or the `elevatedButtonTheme` (or outlined/text/etc) in `ThemeData`, there is no way to allow buttons to continue being platform adaptive AND force them to a specific platform in certain scenarios (like my Widgetbook).

The Material buttons, and other places that use `VisualDensity` should retrieve the target platform from `ThemeData` as the above comment suggests.

## Steps to Reproduce

1. Run a standard `flutter create` app on web or desktop with the following contents of the Scaffold body:
```
  final theme = Theme.of(context);
  return Row(
    children: [
      ElevatedButton(
        onPressed: () {},
        child: const Text('Button'),
      ),
      Theme(
        data: theme.copyWith(platform: TargetPlatform.android),
        child: ElevatedButton(
          onPressed: () {},
          child: const Text('Button'),
        ),
      ),
    ],
  );
```

**Expected results:**
<img width="166" alt="image" src="https://user-images.githubusercontent.com/3476942/228858091-d18812d1-d418-4995-8526-17caaed24095.png">
The first button due to being on web/desktop platform would render with a `compact` density while the second renders with `comfortable` due to the ambient Theme supplying a new `TargetPlatform`

**Actual results:**
<img width="173" alt="image" src="https://user-images.githubusercontent.com/3476942/228858440-4585dadb-d7e3-4108-8ec2-90e55057f5cc.png">
Two buttons that look the same with a `compact` density despite an explicit `TargetPlatform.android` and the knowledge from the documentation that "Widgets from the material library should use [Theme.of] to determine the current platform for styling purposes, rather than using [defaultTargetPlatform]".

**Comment by Calpoog** [comment]
Perhaps a solution involves a new VisualDensity constructor `adaptiveTargetPlatformDensity` which takes and uses a `BuildContext` for looking up the ambient `TargetPlatform` and performing the same logic on that as is already done with `adaptivePlatformDensity`.

**Comment by gspencergoog** [comment]
> Perhaps a solution involves a new VisualDensity constructor adaptiveTargetPlatformDensity which takes and uses a BuildContext for looking up the ambient TargetPlatform and performing the same logic on that as is already done with adaptivePlatformDensity.

I understand why you might ask for this.  However, in order to access the proposed `adaptivePlatformDensity` factory constructor, the caller has to be able to supply a context, which isn't always possible. In particular, there is no access to a context inside of the [`ThemeData` constructor ](https://cs.opensource.google/flutter/flutter/+/master:packages/flutter/lib/src/material/theme_data.dart;l=451;drc=0300cfa60382fe2cb1e154ee98a590996e22d59a) used to create the default theme, which is the reason it doesn't do this already. Adding a required `BuildContext` to the ThemeData constructor would (as you can imagine) be a very large breaking change, and not really in keeping with the usage of that constructor.

Visual density is usually an app-wide setting, which can be overridden for all, or certain sections, of the widget tree by supplying a new theme with a new visual density.

To define the visual density for a subtree, you should be able to do something like this:
```dart
Theme(
  data: Theme.of(context).copyWith(visualDensity: VisualDensity.standard),
    child: /* ... */,
)
```

This will override the density on all platforms to be `standard` (or zero, which is the default on mobile platforms).

> The Material buttons, and other places that use `VisualDensity` should retrieve the target platform from `ThemeData` as the above comment suggests.

They already do this, sort of: they do it by retrieving the visual density from the theme data, not the platform.

**Comment by Calpoog** [comment]
I could just make Widgetbook wrap each of my widgets with an overridden theme, but I'm thinking that I shouldn't need to.

The material widgets when they don't receive an explicit density fall back to `defaultTargetPlatform` for determining which density to use. This ignores the `TargetPlatform` set by the `ThemeData`. What I'd like is to override the platform for a subtree and have all platform-based decisions (like density, but I'm sure there are others) respect that, without having to explicitly override the density (or other properties) that a target platform dictates.

For instance if in the future there are 10 different properties of material widgets which by default depend on the platform, if they make decisions based on the `defaultTargetPlatform` and not the `platform` on `ThemeData`, then in order to:

1. Have my app behave default for each platform
2. AND have a specific subtree overridden to behave like a specific platform

I have to know the 10 different platform-specific properties and override those explicitly rather than simply override the `platform`.

I know that's a bit of a stretch, because here it is simply one property that is affected. Hopefully that larger tradeoff helps illustrate the point. I'd like the `platform` to be the only thing needed, with no extra knowledge of *how* that affects things, to make a subtree behave like that platform.

**Comment by gspencergoog** [comment]
> The material widgets when they don't receive an explicit density fall back to defaultTargetPlatform for determining which density to use.

Where do you see that happening? Because they typically don't. For example, in `ElevatedButton`, it falls back to the `visualDensity` of the ambient theme. See [here](https://cs.opensource.google/flutter/flutter/+/master:packages/flutter/lib/src/material/elevated_button.dart;l=651;drc=598ebfbd66db86558a0e521a57011ee0877df799) for the Material 3 default that explicitly does this, and [here](https://cs.opensource.google/flutter/flutter/+/master:packages/flutter/lib/src/material/elevated_button.dart;l=379;drc=598ebfbd66db86558a0e521a57011ee0877df799) for the Material 2 default that does this. Perhaps there are some widgets which don't conform to that, but I think that's a typical pattern.

> For instance if in the future there are 10 different properties of material widgets which by default depend on the platform, if they make decisions based on the defaultTargetPlatform and not the platform on ThemeData ...

Ahh, I see what you're concerned about.  I think we already typically try to do what you are requesting where we have access to a context.  Do you have some concrete examples where this isn't the case?

**Comment by Calpoog** [comment]
You're correct. But the ambient theme's visualDensity defaults to `platformAdaptiveDensity` which uses a switch case on `defaultTargetPlatform` which is *not* in line with the current `TargetPlatform` `target` of the ambient theme.

[Here](https://cs.opensource.google/flutter/flutter/+/master:packages/flutter/lib/src/material/theme_data.dart;drc=598ebfbd66db86558a0e521a57011ee0877df799;l=449) is where the `ThemeData` does that.

And [here](https://cs.opensource.google/flutter/flutter/+/master:packages/flutter/lib/src/material/theme_data.dart;drc=598ebfbd66db86558a0e521a57011ee0877df799;l=2627) is where that switch statement is.

Elsewhere in that file I can see that the material tap target size also relies on `defaultTargetPlatform` rather than the ambient `TargetPlatform` from `target` on `ThemeData`.

**Comment by gspencergoog** [comment]
Yes, the `ThemeData` doesn't have access to a `BuildContext` from which to draw another default from, so it does use the `defaultTargetPlatform` if you don't supply a `platform` to it.

I do now see that we have a bug, and this sounds like exactly what you were trying to point out, where we initialize most `ThemeData` properties that depend on platform using the theme's own `platform` value, but `visualDensity` is initialized using `adaptivePlatformDensity`, which always uses `defaultTargetPlatform`. We should probably fix that to use the `ThemeData.platform` value, and maybe that would help your situation, at least for the `visualDensity` attribute.

Can you tell me more about how you think it should work?  How do you want to tell the app what parts of your widget tree should be rendered as Android/iOS, and which should be rendered as the platform default? It sounds like you want to specify a `Theme` with a `ThemeData` that has the `platform` set, and have all of the other properties in the `ThemeData` that depend on `platform` recalculate based on that `platform`. Is that about right?

**Comment by gspencergoog** [comment]
I created a PR for the `visualDensity` initialization bug.

**Comment by Calpoog** [comment]
I think exactly that makes sense. Rather than it defaulting to adaptivePlatformDensity, it can instead default (at time of construction) to that same logic but based upon its own `platform` in choosing one of the other densities. That accomplishes what I would need. If I created a new `Theme` giving it a new `ThemeData` with `copyWith` and set the platform explicitly to Android on a Desktop platform for example, the subtree would correctly get the Android specific tap target size and visual density. If I did not explicitly specify a `VisualDensity`, it would take on the characteristics of the `platform` if that was supplied, or fall back to the `defaultTargetPlatform` which is *definitely* the type of platform that it's running on.

**Comment by gspencergoog** [comment]
Okay, great, so once the patch lands, please close this issue if it solves your problem.
